extensions/amp-story-captions/0.1/track-renderer.js

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {listen} from '../../../src/event-helper';
+import {removeChildren, removeElement} from '../../../src/dom';
+import {setStyles} from '../../../src/style';
+import {toArray} from '../../../src/core/types/array';
+
+// Class used for sections of text that is in the future (for ASR-style captions).
+const FUTURE_CUE_SECTION_CLASS = 'amp-story-captions-future';
+
+/**
+ * Parses a WebVTT timestamp and returns the time in seconds from the start of the video.
+ * https://www.w3.org/TR/webvtt1/#webvtt-timestamp
+ * @param {string} timestamp
+ * @return {?number}
+ */
+function parseTimestamp(timestamp) {
+  const match = /^(?:(\d{2,}):)?(\d{2}):(\d{2})\.(\d{3})$/.exec(timestamp);
+  if (!match) {
+    return null;
+  }
+  const hours = match[1] ? parseInt(match[1], 10) : 0;
+  const minutes = parseInt(match[2], 10);
+  const seconds = parseInt(match[3], 10);
+  const milliseconds = parseInt(match[4], 10);
+  return hours * 3600 + minutes * 60 + seconds + milliseconds / 1000;
+}
+
+export class TrackRenderer {
+  /**
+   *
+   * @param {!HTMLVideoElement} video
+   * @param {!TextTrack} track
+   * @param {!Element} container
+   */
+  constructor(video, track, container) {
+    /** @private {!HTMLVideoElement} */
+    this.video_ = video;
+
+    /** @private {?TextTrack} */
+    this.track_ = track;
+
+    /** @private {!Element} */
+    this.element_ = container.ownerDocument.createElement('div');
+    container.appendChild(this.element_);
+
+    /** @private {!Array<number>} */
+    this.cueTimestamps_ = [];
+
+    this.render_();
+    this.cueChangeUnlistener_ = listen(track, 'cuechange', () => {
+      this.render_();
+    });
+    this.timeUpdateUnlistener_ = listen(video, 'timeupdate', () => {
+      this.updateTime_();
+    });
+  }
+
+  /** Cleans up listeners and DOM elements. */
+  dispose() {
+    this.cueChangeUnlistener_();
+    this.timeUpdateUnlistener_();
+    removeElement(this.element_);
+  }
```

Comment:
Nit: I'd also remove the strong references to `this.video_` and `this.track_` that might prevent garbage collection, setting them to `null` should do

Reply:
I think modern browsers should be able to handle this when there are no more references to TrackRenderer itself, but I can add this.